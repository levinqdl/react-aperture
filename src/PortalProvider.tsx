import React from "react"
import SlotContext, { ISlotContext } from "./SlotContext"

const { Provider } = SlotContext

class PortalProvider extends React.Component<{}, ISlotContext> {
  counter: number = 0
  registerSlot = (element: HTMLDivElement) => {
    const key = `s${this.counter++}`
    this.setState(({ slots }) => ({
      slots: {
        ...slots,
        [key]: { element, payload: null, renders: 0 },
      },
    }))
    return key
  }
  removeSlot = (key: string) => {
    this.setState(({ slots }) => {
      const newSlots = { ...slots }
      delete newSlots[key]

      return {
        slots: {
          ...newSlots,
        },
      }
    })
  }
  registerRender = (key: string) => {
    this.setState(({ slots }) => {
      const slot = slots[key]
      if (!slot) {
        return null
      }

      return {
        slots: {
          ...slots,
          [key]: { ...slot, renders: slot.renders + 1 },
        },
      }
    })
    return () => {
      this.setState(({ slots }) => {
        const slot = slots[key]
        if (!slot) {
          return null
        }

        return {
          slots: {
            ...slots,
            [key]: { ...slot, renders: slot.renders - 1 },
          },
        }
      })
    }
  }
  setPayload = (key: string, payload: any) => {
    this.setState(({ slots }) => {
      const slot = slots[key]
      return {
        slots: {
          ...slots,
          [key]: { ...slot, payload },
        },
      }
    })
  }
  state: ISlotContext = {
    slots: {},
    registerSlot: this.registerSlot,
    removeSlot: this.removeSlot,
    setPayload: this.setPayload,
    registerRender: this.registerRender,
  }
  render() {
    const { children } = this.props
    return <Provider value={this.state}>{children}</Provider>
  }
}
export default PortalProvider
