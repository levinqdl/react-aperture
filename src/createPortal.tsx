import React, {
  createRef,
  RefObject,
  SFC,
  ComponentClass,
  CSSProperties,
} from "react"
import ReactDOM from "react-dom"
import SlotContext from "./SlotContext"

type ISlotRender = SFC<{}>

interface SlotProps {
  className?: string
  style?: CSSProperties
}

type ISlot = ComponentClass<SlotProps>

const createPortal: () => [ISlot, ISlotRender] = () => {
  let key: string

  class Slot extends React.Component {
    static contextType = SlotContext
    context: any
    elem: RefObject<HTMLDivElement> = createRef()
    componentDidMount() {
      key = this.context.addSlot(this.elem.current)
    }
    render() {
      return <div ref={this.elem} {...this.props} />
    }
  }

  const SlotRender: ISlotRender = ({ children }) => (
    <SlotContext.Consumer>
      {({ slots }) => {
        const elem = slots[key]
        return elem ? ReactDOM.createPortal(children, elem) : null
      }}
    </SlotContext.Consumer>
  )
  return [Slot, SlotRender]
}

export default createPortal
