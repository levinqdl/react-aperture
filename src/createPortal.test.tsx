import React from "react"
import { render, within } from "@testing-library/react"
import PortalProvider from "./PortalProvider"
import createPortal from "./createPortal"

describe("Slot", () => {
  it("renders a portal", () => {
    const [Slot, Render] = createPortal()
    const { getByText } = render(
      <PortalProvider>
        <div>
          Awesome <Slot />
        </div>
        <Render>Portal Layout</Render>
      </PortalProvider>
    )
    const slotContainer = getByText("Awesome")
    within(slotContainer).getByText("Portal Layout")
  })
  it("reveive style & className", () => {
    const [Slot, Render] = createPortal()
    const { getByText, container } = render(
      <PortalProvider>
        <div>
          Slot Container <Slot className="slot" style={{ display: "inline" }} />
        </div>
        <Render>Styled</Render>
      </PortalProvider>
    )
    const slot = getByText("Styled").parentElement
    expect(slot.classList.contains("slot")).toBeTruthy()
    expect(slot.style).toMatchObject({ display: "inline" })
    expect(container.firstChild).toMatchSnapshot()
  })
  test("pass data from Slot to Render", () => {
    const [Slot, Render] = createPortal()
    const { getByText } = render(
      <PortalProvider>
        <div>
          Awesome <Slot payload={{ data: "payload" }} />
        </div>
        <Render>{(payload: any) => <>Portal Layout {payload.data}</>}</Render>
      </PortalProvider>
    )
    const slotContainer = getByText("Awesome")
    within(slotContainer).getByText("Portal Layout payload")
  })
  it("hides Slot if no Render", () => {
    const [Slot, Render] = createPortal()
    const { getByText } = render(
      <PortalProvider>
        <div>
          Awesome <Slot id="slot" />
        </div>
      </PortalProvider>
    )
    expect(document.querySelector("#slot")).toBeNull()
  })
})
