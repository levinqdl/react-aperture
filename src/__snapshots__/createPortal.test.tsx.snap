// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot reveive style & className 1`] = `
.emotion-0 {
  width: 100vw;
  height: 100vh;
}

<div
  class="emotion-0"
>
  <div>
    Slot Container 
    <div
      class="slot"
      style="display: inline;"
    >
      Styled
    </div>
  </div>
</div>
`;
